firebase-firestore/src/main/java/com/google/firebase/firestore/local/LruGarbageCollector.java

```diff
@@ -91,4 +157,79 @@ int removeTargets(long upperBound, SparseArray<?> activeTargetIds) {
   int removeOrphanedDocuments(long upperBound) {
     return delegate.removeOrphanedDocuments(upperBound);
   }
+
+  Results collect(SparseArray<?> activeTargetIds) {
+    if (params.minBytesThreshold == Params.COLLECTION_DISABLED) {
+      Logger.debug("LruGarbageCollector", "Garbage collection skipped; disabled");
+      return Results.DidNotRun();
+    }
+
+    long cacheSize = getByteSize();
+    if (cacheSize < params.minBytesThreshold) {
+      Logger.debug(
+          "LruGarbageCollector",
+          "Garbage collection skipped; Cache size "
+              + cacheSize
+              + " is lower than threshold "
+              + params.minBytesThreshold);
+      return Results.DidNotRun();
+    } else {
+      return runGarbageCollection(activeTargetIds);
+    }
+  }
+
+  private Results runGarbageCollection(SparseArray<?> liveTargetIds) {
+    long start = System.currentTimeMillis();
+    int sequenceNumbers = calculateQueryCount(params.percentileToCollect);
+    // Cap at the configured max
+    if (sequenceNumbers > params.maximumSequenceNumbersToCollect) {
+      Logger.debug(
+          "LruGarbageCollector",
+          "Capping sequence numbers to collect down to the maximum of "
+              + params.maximumSequenceNumbersToCollect
+              + " from "
+              + sequenceNumbers);
+      sequenceNumbers = params.maximumSequenceNumbersToCollect;
+    }
+    long countedTargets = System.currentTimeMillis();
+
+    long upperBound = nthSequenceNumber(sequenceNumbers);
+    long foundUpperBound = System.currentTimeMillis();
+
+    int numTargetsRemoved = removeTargets(upperBound, liveTargetIds);
+    long removedTargets = System.currentTimeMillis();
+
+    int numDocumentsRemoved = removeOrphanedDocuments(upperBound);
+    long removedDocuments = System.currentTimeMillis();
+
+    // TODO(gsoltis): post-compaction?
+
+    String desc = "LRU Garbage Collection:\n";
+    desc += "\tCounted targets in " + (countedTargets - start) + "ms\n";
+    desc +=
+        "\tDetermined least recently used "
+            + sequenceNumbers
+            + " sequence numbers in "
+            + (foundUpperBound - countedTargets)
+            + "ms\n";
+    desc +=
+        "\tRemoved "
+            + numTargetsRemoved
+            + " targets in "
+            + (removedTargets - foundUpperBound)
+            + "ms\n";
+    desc +=
+        "\tRemoved "
+            + numDocumentsRemoved
+            + " documents in "
+            + (removedDocuments - removedTargets)
+            + "ms\n";
+    desc += "Total Duration: " + (removedDocuments - start) + "ms";
+    Logger.debug("LruGarbageCollector", desc);
+    return new Results(/* hasRun= */ true, sequenceNumbers, numTargetsRemoved, numDocumentsRemoved);
+  }
+
+  long getByteSize() {
+    return delegate.getByteSize();
```

Comment:
Looking at this more, I think this should `LruGarbageCollector.getCurrentCacheSize()`. `Current` answers the question of whether this is the size before or after a GC run.